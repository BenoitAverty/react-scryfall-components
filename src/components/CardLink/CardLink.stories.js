import React from 'react';

import { storiesOf } from '@storybook/react';

import CardLink from '.';

storiesOf('CardLink', module)
  .add('CardLink', () => <CardLink>Urborg, tomb of yawgmoth</CardLink>)
  .add('fuzzy', () => <CardLink>Blac Lotus</CardLink>)
  .add('card not found', () => (
    <CardLink>This card doesn&apos;t exist</CardLink>
  ));
