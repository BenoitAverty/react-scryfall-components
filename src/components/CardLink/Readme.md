Simple link to a card:

```js
<CardLink>Urborg, tomb of yawgmoth</CardLink>
```

Link to a card that doesn't exist:

```js
<CardLink>This card doesn't exist</CardLink>
```