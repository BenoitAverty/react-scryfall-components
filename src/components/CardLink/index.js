import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import axios from '../../axios';

import Card from '../Card';

const css = {
  position: 'relative',
  display: 'inline-block',
  '& a': {
    'text-decoration': 'underline dotted',
  },
  '.tooltip': {
    'text-align': 'center',
    width: '146px',
    height: '204px',
    position: 'absolute',
    'background-color': '#ddd',
    bottom: '160%',
    left: '50%',
    'margin-left': '-73px',
  },
  '.tooltip::after': {
    content: '" "',
    position: 'absolute',
    top: '100%' /* At the bottom of the tooltip */,
    left: '50%',
    'margin-left': '-5px',
    'border-width': '5px',
    'border-style': 'solid',
    'border-color': 'black transparent transparent transparent',
  },
};
const CardLinkContainer = styled('div')(css);

class CardLink extends React.Component {
  static propTypes = {
    /**
     * Name of the card. The link will be to the scryfall page of the card, or if
     * it can't be found, to the search page with this string as the query.
     */
    children: PropTypes.string.isRequired,
  };

  constructor(props) {
    super(props);

    this.showTooltip = this.showTooltip.bind(this);
    this.hideTooltip = this.hideTooltip.bind(this);

    this.state = {
      scryfallUri: null,
      showTooltip: false,
    };
  }

  componentDidMount() {
    const { children: cardName } = this.props;

    axios.get(`/cards/named?exact=${cardName}`).then(resp => {
      this.setState({
        scryfallUri: resp.data.scryfall_uri,
        scryfallId: resp.data.id,
      });
    });
  }

  showTooltip() {
    this.setState({ showTooltip: true });
  }

  hideTooltip() {
    this.setState({ showTooltip: false });
  }

  render() {
    const { scryfallUri, scryfallId, showTooltip } = this.state;
    const { children } = this.props;

    const href = scryfallUri || `https://scryfall.com/search?q=${children}`;

    const tooltip =
      showTooltip && scryfallId ? (
        <span className="tooltip">
          <Card id={scryfallId} size={Card.SIZE_SMALL} />
        </span>
      ) : null;

    return (
      <CardLinkContainer
        onMouseOver={this.showTooltip}
        onFocus={this.showTooltip}
        onMouseLeave={this.hideTooltip}
        onBlur={this.hideTooltip}
      >
        {tooltip}
        <a href={href}>{children}</a>
      </CardLinkContainer>
    );
  }
}

export default CardLink;
