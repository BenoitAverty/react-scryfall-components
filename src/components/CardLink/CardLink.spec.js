// Import all that is needed for testing
import React from 'react';
import {
  render,
  cleanup,
  wait,
  fireEvent,
  waitForElement,
} from 'react-testing-library';
import 'jest-dom/extend-expect';

import {
  loadAndApplyFixtures,
  saveFixtures,
} from '../../../tests/axiosMocking';

// Import the component under test
import CardLink from '.';

describe('CardLink component', () => {
  // Setup mocking of axios requests.
  let fixtures;
  beforeAll(async () => {
    fixtures = await loadAndApplyFixtures('./src/components/CardLink');
  });
  afterAll(() => {
    saveFixtures('./src/components/CardLink', fixtures);
  });

  beforeEach(() => {
    cleanup();
  });

  describe('With card name', () => {
    test('snapshot', () => {
      const { container } = render(<CardLink>Reveillark</CardLink>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders a link to the scryfall search', () => {
      const { container } = render(<CardLink>Black Lotus</CardLink>);

      expect(container.querySelectorAll('a')).toHaveLength(1);
      expect(container.querySelector('a')).toHaveAttribute(
        'href',
        'https://scryfall.com/search?q=Black Lotus',
      );
    });

    it('renders a link to the scryfall page after the search is resolved', async () => {
      const { container } = render(<CardLink>Black Lotus</CardLink>);

      await wait(() => {
        expect(container.querySelector('a')).toHaveAttribute(
          'href',
          'https://scryfall.com/card/vma/4/black-lotus?utm_source=api',
        );
      });
    });
  });

  describe('Card Tooltip', () => {
    test('snapshot', async () => {
      const { container, getByText, getByAltText } = render(
        <CardLink>crucible of worlds</CardLink>,
      );

      fireEvent.mouseOver(getByText('crucible of worlds'));
      await waitForElement(() => getByAltText('Crucible of Worlds'));
      expect(container.firstChild).toMatchSnapshot();
    });
    it('shows a card on mouse over', async () => {
      const { container, getByText } = render(
        <CardLink>crucible of worlds</CardLink>,
      );

      fireEvent.mouseOver(getByText('crucible of worlds'));
      await wait(() => {
        expect(container.querySelector('img')).toHaveAttribute(
          'alt',
          'Crucible of Worlds',
        );
      });
    });

    it('hides the card on mouse out', async () => {
      const { container, getByText, getByAltText } = render(
        <CardLink>crucible of worlds</CardLink>,
      );

      // mouse enter, Wait for the image to appear then mouse leave
      fireEvent.mouseOver(getByText('crucible of worlds'));
      await waitForElement(() => getByAltText('Crucible of Worlds'));
      fireEvent.mouseOut(getByText('crucible of worlds'));

      await wait(() => {
        expect(container.querySelector('img')).toBeNull();
      });
    });
  });
});
