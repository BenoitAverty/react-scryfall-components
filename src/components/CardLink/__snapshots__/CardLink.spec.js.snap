// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardLink component Card Tooltip snapshot 1`] = `
<div
  class="css-h863mz-CardLinkContainer e130gayk0"
>
  <span
    class="tooltip"
  >
    <img
      alt="Crucible of Worlds"
      height="680"
      src="https://img.scryfall.com/cards/normal/front/e/b/eb28b35c-28a5-4042-b21d-6d43658a16eb.jpg?1562304863"
      style="display: block;"
      width="488"
    />
  </span>
  <a
    href="https://scryfall.com/card/m19/229/crucible-of-worlds?utm_source=api"
  >
    crucible of worlds
  </a>
</div>
`;

exports[`CardLink component With card name snapshot 1`] = `
<div
  class="css-h863mz-CardLinkContainer e130gayk0"
>
  <span
    class="tooltip hide"
  >
    <img
      alt="Reveillark"
      height="680"
      src="https://img.scryfall.com/cards/normal/front/4/b/4bcf9b52-d249-41e7-8ed1-ba9acc816f30.jpg?1547515787"
      style="display: block;"
      width="488"
    />
  </span>
  <a
    href="https://scryfall.com/card/uma/31/reveillark?utm_source=api"
  >
    Reveillark
  </a>
</div>
`;
